sql/core/src/main/scala/org/apache/spark/sql/execution/joins/MergeAsOfJoinExec.scala

```diff
@@ -85,31 +92,38 @@ case class MergeAsOfJoinExec(
     var rPrev = rHead.copy()
 
     currLeft._2.map(lHead => {
-        breakable {
-          while (rHead.getInt(0) <= lHead.getInt(0)) {
-            // TODO make index agnostic and check by type (timestamp)
-            var rHeadCopy = rHead.copy()
-            if (currRight._2.hasNext) {
-              rPrev = rHeadCopy.copy()
-              rHeadCopy = currRight._2.next()
-            } else {
-              break
-            }
+      val leftKeyProj = UnsafeProjection.create(Seq(leftOn), left.output)
```

Comment:
You can lift the creation of projection out side the loop. You only need one projection per zipped partitions